feat(renderer): adding a first implementation of Labels

zarov opened the pull request:
This is a first PR in managing labels (either text, icon or both) in iTowns. There is still a lot of room for improvement, but I would like to see this first bit going into the codebase before it is too big to review. Below is the result for the published `vector_tile_raster_2d` example.

![image](https://user-images.githubusercontent.com/741255/72812206-5c85fa80-3c61-11ea-9ba7-f935ee0deafb.png)

Here is the result for another unpublished example using vector tile as well:

![image](https://user-images.githubusercontent.com/741255/72813094-1336aa80-3c63-11ea-929c-68b8e0a4f84c.png)

#### feat(renderer): add a renderer for Labels

This new renderer (Label2DRenderer) is launched in c3DEngine, alongside
WebGLRenderer. It manages Labels present on the scene.

#### feat(label): add a first implementation of Labels

This commit has two main aspects:
- `Label` object, which represents a text and/or an icon. It has a 3D
position, and it is displayed through DOM.
- `LabelLayer`, which is managing the present labels on a scene. It can
be created on its own, or it can get data from another layer.

There has also been some update in the `vector_tile_raster_2d` example,
as well as in the parsing of vector tiles and style.

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    render(scene, camera) {
+        viewMatrix.copy(camera.matrixWorldInverse);
```
why you copy `camera.matrixWorldInverse`? `viewMatrix` is useless?

zarov replied:
It is true, I'll fix it

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,141 @@
+        //          - hide its children Label
+        //          - iterate through other children (essentially TileMesh)
+        //      - otherwise iterate through its children
+        // - it not visible, hide all its potential Label children
```
May be, the culling could be move in update `LabelLayer`.

It's possible to group Labels in one `div`, if it's in same `children` array?

Where is the case where `style.display = ''` ?

gchoqueux replied:
do you group the labels in a div?

zarov replied:
Yes, this is done in `LabelLayer`

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,141 @@
+        // each Label is stored in labels to be sorted in zOrder later
+        labels.push(object);
+
+        if (object.content.parentNode == null) {
```
what is this part?

zarov replied:
The label is attached to the tile, but it hasn't been added to the DOM yet.

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,141 @@
+     *   are overlapping
+     * - move the remaining labels to their correct positions
+     */
+    zOrder() {
```
There are 3 loops (4 with `labelLayer.update()`), how to minimize tree paths?

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,141 @@
+            label = labels[i];
+            label.content.style.zIndex = l - i;
+            label.content.style.display = '';
+            label.content.style[this.transform] = `${label.anchor} translate(${label.x * this.halfWidth + this.halfWidth}px,${-label.y * this.halfHeight + this.halfHeight}px)`;
```
from what you've told me, this transformation is the most consuming...?

zarov replied:
Yes, but by correctly culling in the process before, we should be able to keep it quite fast (I think like a maximum of 100~ labels updated each time)

gchoqueux reviewed (commented):
I comment only the first commit

gchoqueux commented:
Could move label example to gpx raster ? as this's not yet sufficiently optimized for the vector tiles.

zarov committed:
feat(renderer): add a renderer for Labels

This new renderer (Label2DRenderer) is launched in c3DEngine, alongside
WebGLRenderer. It manages Labels present on the scene.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+     * it can be accessed through `label.content.style`, but it is highly
+     * discouraged to do so.
+     */
+    constructor(content, position, config = {}, style) {
```
It's better to support only `Coordinates` to place label.

zarov replied:
Yes, this is a residue of my first work on this, when I did not know any coordinate to place a label. I'll remove it.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+
+        this.isLabel = true;
+
+        if (position.isVector3) {
```
Which interest to support `Vector3`?

zarov replied:
See other comment

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+            throw new Error('position is not valid; must be a THREE.Vector3 or Coordinates');
+        }
+
+        this.boundaries = { left: 0, right: 0, top: 0, bottom: 0 };
```
May be use `Box2`?

zarov replied:
I'll try it. I'm not sure about the gain here, as it introduces a heavier structure than this simple object.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+            this.content.appendChild(content);
+        }
+
+        if (config.id) {
```
What is this `id`?

zarov replied:
As labels are HTML node, this corresponds to the `id` attribute of a node.

zarov replied:
I removed it

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+        if (config.id) {
+            this.content.id = config.id;
+        }
+        if (config.class) {
```
What is this `class`?

zarov replied:
Same as `id`.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+            this.zOrder = style.text.zOrder;
+        }
+
+        if (style && style.icon) {
```
Could you write a function `Style.toCss()`?

zarov replied:
Good idea !

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+     *
+     * @return {boolean} `true` if in, otherwise `false`.
+     */
+    isInNode(node) {
```
It's too specific function, no?

zarov replied:
Maybe, it was to make things more understandable in `LabelLayer`. I'll change it to see how it looks.

zarov replied:
I removed it

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+     * correct hiding of the label. The style of the content of the label should
+     * not be modified directly to hide it.
+     */
+    hide() {
```
Could you write a `Label.visible` setter to encapsulate the display `content`.

zarov replied:
Ok

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+     * `transfom`, but can `-webkit-transform` for example, depending on the
+     * browser.
+     */
+    move(x, y, transform) {
```
may be use a global constant for `transform` property ?

zarov replied:
It makes sense yes, I'll change it.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+            this.x = x;
+            this.y = y;
+            this.content.style[transform] = `${this.anchor} translate(${x}px, ${y}px)`;
+            this.updateBoundaries();
```
it's necessary to `updateBoundaries` only if the label is tested in the Grid?

zarov replied:
A label is only moved if added to the Grid, and then we need boundaries to test it, so it won't change much.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+     *
+     * @return {boolean} `true` if intersecting, `false` otherwise.
+     */
+    intersection(label) {
```
intersection isn't used?

zarov replied:
Not anymore, I'll remove it

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+    updateDimensions() {
+        if (!this.width && !this.height) {
+            rect = this.content.getBoundingClientRect();
+            this.width = rect.width;
```
may be write getter to link this.width and rect.width? or maybe it's too costly?

zarov replied:
It is too costly, as we access `width/height` a multiple times, and we don't want to check or recompute them everytime. Also, they shouldn't change much, once they are set they should be good.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+    getAnchorPosition(anchor) {
```
could you move to `Style.toCss()`

zarov replied:
Yes

gchoqueux commented on src/Core/TileMesh.js:
```diff
@@ -41,6 +41,18 @@ class TileMesh extends THREE.Mesh {
+
+        // Necessary event listener, to remove any child (beside OBB) attached
+        // to this tile
+        this._removedHandler = this.addEventListener('removed', () => {
```
Could you explain this part? there's `ObjectRemovalHelper`, it is similar ?

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,166 @@
+                    g.label.config.crs = g.label.config.crs || this.crs;
+
+                    if (g.label.position.z == Infinity) {
+                        g.label.position.z = DEMUtils.getElevationValueAt(this.parent, g.label.position) + 100;
```
the z position must be updated with the parent tile, in `LabelLayer.update` function.

zarov replied:
Done, thanks

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,166 @@
+            f.geometry.forEach((g) => {
+                if (g.label && g.label.content && g.label.position) {
+                    g.label.config = g.label.config || {};
+                    g.label.config.crs = g.label.config.crs || this.crs;
```
why save `config` and `crs` in geometry?
`crs` is already in Feature.
Pass directly `params` in Label constructor?

zarov replied:
I removed the complicated `config` object for now. I need to add back `minlevel` and `maxlevel`.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,166 @@
+                        label.coordinates.x = (label.coordinates.x / data.scale.x) - data.translation.x;
+                        label.coordinates.y = (label.coordinates.y / data.scale.y) - data.translation.y;
+
+                        label.coordinates.as(this.crs, label.coordinates);
```
it would be best to keep the coordinates in their original projection. Especially when updating the altitude?

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,166 @@
+                        label.coordinates.toVector3(label.position);
+
+                        // The label needs a translation to a local reference
+                        label.needsMoveToLocal = true;
```
`needsMoveToLocal` isn't necessary? we can know when we add the label to local object

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,166 @@
+            return;
+        }
+
+        const extentsDestination = node.getExtentsByProjection(this.source.projection) || [node.extent];
```
`[node.extent]` is necessary?

gchoqueux replied:
WDYT?

zarov replied:
No idea, I shamelessly copied this from `FeatureProcessing`.

gchoqueux replied:
to verify

gchoqueux replied:
it's not necessary in `featureProcessing`, could you try it?

zarov replied:
I removed it

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,166 @@
+                const labelsToAdd = [];
+
+                labels.forEach((label) => {
+                    if (label.parent || !label.isInNode(node)) { return; }
```
* `!label.isInNode(node)` : why labels aren't in node extent  ?
    you could add filtering to remove label  that are out of extents tile.

If label have parent, his `layerUpdateState` should keep him from returning here.

zarov replied:
This is for `FileSource`, as each Label is attached to a tile. If this check is not done,  there is a duplication of label. But maybe it can (and should) be done before. I'm looking into it.

gchoqueux replied:
`isInNode `is removed

gchoqueux commented on src/Parser/GeoJsonParser.js:
```diff
@@ -51,6 +51,15 @@ const toFeature = {
         geometry.properties.style = new Style().setFromGeojsonProperties(properties, feature.type);
         this.populateGeometry(crsIn, coordsIn, geometry, setAltitude, feature);
+
+        // TODO: take a decision for labels with other FEATURE_TYPES
```
We should keep `FEATURE_TYPES.POINT` and have a convert `FeaturePoint2Label`

zarov replied:
I moved the parts in parsers to `LabelLayer#convert`.

gchoqueux commented on src/Parser/GeoJsonParser.js:
```diff
@@ -42,7 +42,7 @@ const toFeature = {
         geometry.updateExtent();
     },
-    default(feature, crsIn, coordsIn, filteringExtent, setAltitude, properties) {
+    default(feature, crsIn, coordsIn, filteringExtent, setAltitude, properties, style) {
```
why inherit the style?

zarov replied:
I removed this part

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,163 @@
+        // culling
+        this.culling(scene);
+
+        this.grid.forEach(c => c && c.show());
```
Could move `Label.show()` when it inserts in the grid?

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,163 @@
+
+            if (object.material && !object.material.visible) {
+                object.children.forEach((c) => {
+                    if (c.isLabel) {
```
may be group label in `THREE.Group` to avoid loop on all children.

gchoqueux replied:
add event in `material.visible` to hidden/show the label.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,140 @@
+
+        if (style) {
+            this.anchor = style.getTextAnchorPositionInCSS();
+            this.zOrder = style.text.zOrder;
```
could you clean and comment this part?

zarov replied:
Done

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,140 @@
+            this.zOrder = style.text.zOrder;
+
+            style.addIconToElement(this.content);
+            style.applyToHTML(this.content);
```
could you merge the functions?

zarov replied:
Done in #1345

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+            f.geometry.forEach((g) => {
+                const geometryField = g.properties.style && g.properties.style.text.field;
+
+                if (!geometryField && !featureField && !layerField) {
```
`Style` handling needs architecture refactoring.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+
+                // NOTE: this only works because only POINT is supported, it
+                // needs more work for LINE and POLYGON
+                vertices = f.getVerticesFromGeometry(g);
```
Could you strike directly the point?

zarov replied:
Done

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+                    label.needsMoveToLocal = true;
+                }
+
+                if (f.size == 2) {
+                    label.needsAltitude = true;
+                }
```
Why ?

zarov replied:
I don't see another way to do it ? Keep in mind that not all labels need have a computed altitude.

gchoqueux replied:
ok

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+
+    // TODO: see how to factorize this behavior with FeatureProcessing which is
+    // similar.
+    update(context, layer, node, parent) {
```
Very similar to `featureProcessing#update`?

zarov replied:
The part before the `scheduler.execute`, but I can discard the comment as it was more of something I thought at when I began to work on it.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+                const labelsToAdd = [];
+
+                labels.forEach((label) => {
+                    if (label.parent || !node.extent.isPointInside(label.coordinates)) { return; }
```
it is necessary to see the best place to filter, there is already a mechanism in the geojson parser

zarov replied:
If I don't do this, I have a problem with `FileSource`. As it call `update` for each tile, and for a file such as the example with GPX, it always read __all__ points, and thus it creates a `Label` each time. So we have at the end a lot of duplicated labels. If you have another solution, I'm happy to implement it !

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+
+                        if (l.needsAltitude) {
+                            // TODO: add a Style property for the offset
+                            l.position.z = DEMUtils.getElevationValueAt(this.parent, l.coordinates) + 2;
```
Update altitude label when the DEM elevation is updated.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+                    labelsToAdd.push(label);
+                });
+
+                if (!node.parent) {
```
move before `labelsToAdd` building? WDYT?

zarov replied:
Yes, done

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+                        }
+
+                        l.visible = true;
+                        l.updateDimensions();
```
not necessary?

zarov replied:
I have to do it after adding it to the DOM, I can't really see another place to do it, because I can get the dimensions only when the content of the Label is visible.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,197 @@
+                }
+            });
+
+            node.updateMatrixWorld();
```
not necessary, update only label.

zarov replied:
Yes, done

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,156 @@
+        // culling
+        this.culling(scene);
+
+        this.grid.forEach((c) => {
```
move in `grid#insert`

zarov replied:
Forgot to do it, done !

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,138 @@
+
+        this.boundaries = { left: 0, right: 0, top: 0, bottom: 0 };
+
+        this.content = document.createElement('div');
```
It's possible to create css file and import properties to instance the div?

zarov replied:
Yes, I'll do it in another PR, as it can be done entirely in `Style`.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,138 @@
+
+        this.addEventListener('removed', () => { this.visible = false; });
+
+        // FIXME: this needs to be moved outside, to be executed only on start.
```
Ok.
There's the same `document` in `FeatureToTexture.js` there is no problem.

zarov replied:
Indeed, that's weird. I'm looking into it.

zarov replied:
It seems that fixing other things (having nothing to do with this) made this works outside the constructor. I'll check a few things before pushing it.

gchoqueux replied:
Fixed with `--file bootstrap.js`

gchoqueux commented on src/Core/TileMesh.js:
```diff
@@ -41,6 +41,18 @@ class TileMesh extends THREE.Mesh {
+
+        // Necessary event listener, to remove any child (beside OBB) attached
+        // to this tile
+        this._removedHandler = this.addEventListener('removed', () => {
```
Could you move this handler, when the labels are added to the parent?
If you handle when Labels are added, you could remove the test `this.children.length > 1` because there will labels in children.

zarov replied:
Good idea, will do !

gchoqueux commented on src/Core/TileMesh.js:
```diff
@@ -41,6 +41,18 @@ class TileMesh extends THREE.Mesh {
+            // > 1 permits to filter tiles with only OBB
+            if (this.children.length > 1) {
+                const toRemove = [];
+                this.children.forEach((c) => { if (c.isLabel) { toRemove.push(c); } });
```
This code could work?
`this.children.forEach((c) => { if (c.isLabel) { this.remove(c); } });`

zarov replied:
No, it breaks the `forEach` unfortunately.

gchoqueux replied:
I tested without problem, what's breacking message ?

zarov replied:
It is because of this line https://github.com/mrdoob/three.js/blob/master/src/core/Object3D.js#L371

By calling `splice` in a `forEach`, the array is completely rearranged, but the `forEach` doesn't know it, so you miss a test on one element. If you try this, you can see what I mean:
```js
const b = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
// this should remove 3 and 4 from the array
b.forEach((e, i, a) => { if (e == 3 || e == 4) { a.splice(i, 1); } });
// but if you print b, you'll see that 4 is still present
console.log(b);
```

gchoqueux replied:
ok

zarov replied:
As I moved it in `LabelLayer`, I was able to remove the necessary selection of labels (see here https://github.com/iTowns/itowns/pull/1303/files#diff-9229f99ccb02f6c4aa8d3d42de700a92R179-R182)

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,189 @@
+                // NOTE: the config param is empty, but will be filled in the
+                // next version.
+                const label = new Label(content, coord.clone(), g.properties.style || f.style || this.style);
+                label.layer = label;
```
it isn't `label.layer = layer;`?

If it's the case, could add layer in  `constructor parameter `?

zarov replied:
This is a mistake indeed, thanks. I copied this from `FeatureProcessing`, is it really necessary ? I can't remember the use for it.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,189 @@
+                // coord.y = (coord.y / data.scale.y) - data.translation.y;
+                coord.as(this.crs, coord);
+                coord.toVector3(label.position);
+                label.needsToBeMoved = true;
```
why don't do this in constructor?

```suggestion
      // need to add layer parameter in Label constructor
      this.coord.as(this.layer.crs, this.position);
      this.needsToBeMoved = true;
```

zarov replied:
This won't work,  `Coordinates#as()` can't take a `THREE.Vector3` as second argument. But the operation could be done in the constructor indeed, I'm trying it.

zarov replied:
This is done in the constructor now

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,189 @@
+                    if (label.needsToBeMoved) {
+                        node.worldToLocal(label.position);
+                        label.needsToBeMoved = false;
+                    }
```
to remove `needsToBeMoved` 

```suggestion
if (label.needsAltitude) {
    // TODO: add a Style property for the offset
    label.coordinates.z = DEMUtils.getElevationValueAt(this.parent, label.coordinates);
}
label.coord.as(layer.crs, label.position);
node.worldToLocal(label.position);
```

zarov replied:
It won't work, as `worldToLocal` will be called everytime we add back a `Label` to the scene. Also, it applies a conversion everytime too.

gchoqueux replied:
I'm not sure because label is added to node, only one time, isn't it?

zarov replied:
Not if we unzoom and the node is deleted, and we zoom again, the node is added back and we go again here.

gchoqueux replied:
but if the node is deleted a new label is instanced, no?

zarov replied:
No, it is only hidden. In my first thoughts, it was costing less to keep a hidden label instead of creating one everytime. But if there is a lot of them, it can be costly indeed. I'll think of a solution, but it seems that adding a `removed` handler will do the trick.

zarov replied:
In fact, as it is parsed data, it automatically goes into cache (see `DataSourceProvider`). So I'm not sure on what to do.

gchoqueux replied:
Brainstorm a better solution

gchoqueux replied:
we're putting aside for the moment

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,189 @@
+
+                    if (label.needsAltitude) {
+                        // TODO: add a Style property for the offset
+                        label.position.z = DEMUtils.getElevationValueAt(this.parent, label.coordinates) + 2;
```
you add `+2` to world coordinate and not to altitude.

zarov replied:
At this point, the label has been moved to a local position (so its parent is a node). The orientation of `z` is following the altitude axis.

gchoqueux replied:
yes, sorry!

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,189 @@
+
+                    label.visible = true;
+                    label.updateDimensions();
+                    label.updateMatrixWorld();
```
May be override `Label#updateMatrixWorld` to encapsulate `updateDimensions`?

zarov replied:
Won't `updateDimensions` be called too many times if I do this ?

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,151 @@
+            if (object.material && !object.material.visible) {
+                object.children.forEach((c) => {
+                    if (c.isLabel) {
+                        c.visible = false;
```
could you remove `c.visible = false;`? because is already hidden in `computeLabelPosition`

zarov replied:
But `computeLabelPosition` isn't called for every label, so it isn't always hidden.

gchoqueux replied:
Add event in `material.visible` to hidden/show the label.

zarov replied:
Added it in #1351

zarov committed:
feat(label): add a first implementation of Labels

This commit has two main aspects:

- Label object, which represents a text and/or an icon. It has a 3D
  position, and it is displayed through DOM.
- LabelLayer, which is managing the present labels on a scene. It can be
  created on its own, or it can get data from another layer.

zarov commented:
I rebased and cleaned everything, it should be good to go now !

zarov commented:
Rebased again, on but needs to wait for #1353 to be merged before.

gmaillet commented:
I think that there is confusion in updateElevationFromLayer between z in the geocentric coordinates and z above the ground. 
It can produce large errors in the placement of labels.

gmaillet commented:
In addition to the altitude issue (see above), there are errors related to updating the position in local coordinates to the tile (worldToLocal). For exemple, when a tile is removed, label survive and could be added in a new tile. In these cases,  position need to be updated.
I propose to separate the coordinated part (in the Label class) and the position part in the LabelLayer class.

```diff
diff --git a/src/Core/Label.js b/src/Core/Label.js
index 6f5848ba..0ea8e826 100644
--- a/src/Core/Label.js
+++ b/src/Core/Label.js
@@ -75,15 +75,8 @@ class Label extends THREE.Object3D {
         if (options.translation && options.scale) {
             this.coordinates.x = (this.coordinates.x / options.scale.x) - options.translation.x;
             this.coordinates.y = (this.coordinates.y / options.scale.y) - options.translation.y;
-            this.coordinates.as(options.crs, coord);
-        } else {
-            this.coordinates.as(options.crs || this.coordinates.crs, coord);
         }
 
-        coord.toVector3(this.position);
-
-        this.needsToBeMoved = true;
-
         this.boundaries = { left: 0, right: 0, top: 0, bottom: 0 };
 
         this.content = document.createElement('div');
@@ -139,7 +132,11 @@ class Label extends THREE.Object3D {
     }
 
     updateElevationFromLayer(layer) {
-        this.position.z = DEMUtils.getElevationValueAt(layer, this.coordinates, DEMUtils.FAST_READ_Z) || 0;
+        const alti = DEMUtils.getElevationValueAt(layer, this.coordinates, DEMUtils.FAST_READ_Z);
+        if (alti && alti > -99999 && alti != this.coordinates.z) {
+            this.coordinates.z = alti;
+            return true;
+        }
     }
 }
 
diff --git a/src/Layer/LabelLayer.js b/src/Layer/LabelLayer.js
index fcd7a83d..6aed63ab 100644
--- a/src/Layer/LabelLayer.js
+++ b/src/Layer/LabelLayer.js
@@ -133,7 +133,12 @@ class LabelLayer extends Layer {
         if (!node.layerUpdateState[node.material.getElevationLayer().id].canTryUpdate()) {
             node.children.forEach((c) => {
                 if (c.isLabel && c.needsAltitude) {
-                    c.updateElevationFromLayer(this.parent);
+                    if (c.updateElevationFromLayer(this.parent)) {
+                        c.coordinates.as(this.crs, coord);
+                        coord.toVector3(c.position);
+                        node.worldToLocal(c.position);
+                        c.updateMatrixWorld();
+                    }
                 }
             });
         }
@@ -183,14 +188,13 @@ class LabelLayer extends Layer {
 
                     label.minzoom = node.extent.zoom;
 
-                    if (label.needsToBeMoved) {
-                        node.worldToLocal(label.position);
-                        label.needsToBeMoved = false;
-                    }
 
                     if (label.needsAltitude) {
                         label.updateElevationFromLayer(this.parent);
                     }
+                    label.coordinates.as(this.crs, coord);
+                    coord.toVector3(label.position);
+                    node.worldToLocal(label.position);
 
                     node.add(label);
                     label.updateMatrixWorld();

```

zarov commented:
Thanks a lot, it seems to solve all my problems ! I'm implementing it right now

gchoqueux commented on examples/source_file_gpx_raster.html:
```diff
@@ -59,29 +60,42 @@
+                            transform: 'uppercase',
+                            font: ['Arial', 'sans-serif'],
+                            halo: {
+                                color: 'white',
```
There's not halo in the gpx example?

zarov replied:
There is not anymore, because I can't merge all the styles unfortunately, so I have to choose a style among three: https://github.com/iTowns/itowns/pull/1303/files/9322d13d0f4a2fdf548bcc7704f684eec88d151e#diff-9229f99ccb02f6c4aa8d3d42de700a92R102
I am not sure how I can fix this problem yet.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * @param {Element|string} content - The content; can be a
+     * string, with or without HTML tags in it, or it can be an Element.
+     * @param {Coordinates} coordinates - The world coordinates, where to place
```
What is the coordinates projection? it's the same of `view.referenceCrs`?

zarov replied:
Not necessary, it can be anything, as `label.position` is not affected by the coordinates projection.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+        this.isLabel = true;
+        this.coordinates = coordinates;
+
+        if (options.translation && options.scale) {
```
Could you move the transformation out of the constructor?
You could add a method in `Feature` to get the good coordinates.

zarov replied:
Ok, will do

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+
+        coord.toVector3(this.position);
+
+        this.needsToBeMoved = true;
```
boolean to move on local object?
Could you add a comment?

zarov replied:
This has been removed, I'm waiting on #1377 to be merged to rebase this PR.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+     */
+    preMove(x, y) {
+        if (x != this.x || y != this.y) {
+            this.x = x;
```
could you initialize `x` and `y` in constructor in object `projectedPosition`, so as not to confuse with `label` 3d position?

zarov replied:
Done !

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    move() {
+        this.visible = true;
```
I'm not sure to display the label when the method is `move`. 
Even if it's the rendering loop they're done in sequence.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,224 @@
+                }
+
+                labels.forEach((label) => {
+                    if (label.parent) { return; }
```
this can't happen?
I kind of tested label has no parent here.

zarov replied:
Not anymore indeed, I'll remove it.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+        }
+    }
+
+    updatePositionFromNode(node, crs) {
```
Could you add property `layer` in label and remove crs param

```js
update3DPosition(crs) {
      this.coordinates.as(crs, coord);
      coord.toVector3(this.position);
      if (this.parent) {
            this.parent.worldToLocal(this.position);
      }
      this.updateMatrixWorld();
}
```

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,224 @@
+                });
+            });
+
+            if (labelsDiv.length > 0) {
```
because  `if (label.parent) { return; }` could be removed so `labelsDiv` could be remove?

zarov replied:
No, because `result` can contain empty arrays.

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,146 @@
+     * `getBoundingClientRect`.  It updates `width` and `height` property of the
+     * label, and the boundaries.
+     */
+    updateDimensions() {
```
could you rename `initDimensions `because is called only one time?

zarov replied:
Yes done

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,224 @@
+                // Batch update the dimensions of labels all at once to avoid
+                // redraw for at least this tile.
+                result.forEach(labels => labels.forEach(label => label.updateDimensions()));
+                result.forEach(labels => labels.forEach((label) => { label.visible = false; }));
```
There are 2 loops for performance?

Could you try :
```js
result.forEach(labels => labels.forEach((label) => {
    label.updateDimensions();
    label.visible = false;
```
or 
```js
result.forEach(labels => labels.forEach((label) => {
    label.content.style.display = 'hidden';
    label.updateDimensions();
    label.content.style.display = 'none';
}));
```

zarov replied:
No, this needs to be separated for performance: it is faster to batch the reads (here `updateDimensions`) and then batch the `writes` (here `display = 'none'`).

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,224 @@
+
+                // Reorganize all Label to be the first in the list of children,
+                // to have an horizontal rendering instead of a vertical one
+                node.children.sort(c => (c.isLabel ? -1 : 1));
```
I don't know `horizontal rendering`

zarov replied:
I removed that part, it was useful before but it doesn't seem the case anymore.

gchoqueux commented on src/Renderer/Label2DRenderer.js:
```diff
@@ -0,0 +1,183 @@
+            object.children.forEach(c => this.culling(c, currentMaxZoom, extent));
+        // By verifying the maxzoom and the presence of the label inside the
+        // visible extent, we can filter more labels.
+        } else if (object.maxzoom <= currentMaxZoom || !extent.isPointInside(object.coordinates)) {
```
the object visibility and material visibility aren't enough to filter out the labels?
may be check `tile.material.visible` and if the children tile have label.

zarov replied:
No, it isn't: a material not visible does not mean a label is not visible anymore.

zarov committed:
example: change the raster GPX example to support labels

gchoqueux commented on src/Core/Feature.js:
```diff
@@ -332,4 +332,18 @@ export class FeatureCollection {
+     * @return {Coordinates} The same coordinates, with transformation applied.
+     */
+    transformCoordinates(coordinates) {
+        coordinates.x = (coordinates.x / this.scale.x) - this.translation.x;
```
Could you add z/altitude transformation ?

zarov replied:
Done, I thought they were `Vector2`

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,154 @@
+        }
+    }
+
+    move() {
```
Could you rename to `updateCssPosition` ?

zarov replied:
Done

gchoqueux commented on src/Core/Label.js:
```diff
@@ -0,0 +1,154 @@
+     * @param {number} x - X coordinates in pixels, from left.
+     * @param {number} y - Y coordinates in pixels, from top.
+     */
+    preMove(x, y) {
```
Could you rename to `updateProjectedPosition`?

zarov replied:
Done

gchoqueux commented on src/Core/View.js:
```diff
@@ -64,10 +63,21 @@ function _preprocessLayer(view, layer, parentLayer) {
+        layer.projection = parentLayer && parentLayer.extent.crs;
+    }
+
+    if (layer.labelEnabled && source) {
```
I'm not sure, there are Layers without source?

zarov replied:
Not anymore, thanks !

gchoqueux commented on src/Core/View.js:
```diff
@@ -64,10 +63,21 @@ function _preprocessLayer(view, layer, parentLayer) {
+        labelLayer.source = source;
+        labelLayer.style = layer.style;
+
+        layer.whenReady = layer.whenReady.then(() => {
+            view.addLayer(labelLayer);
+            return layer;
+        });
```
could you try `view.addLayer(labelLayer);` only?

zarov replied:
It works, thanks !

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,221 @@
+            const featureField = f.style && f.style.text.field;
+
+            f.geometry.forEach((g) => {
+                const minzoom = (g.properties.style && g.properties.style.zoom && g.properties.style.zoom.min)
```
`Style.zoom == undefined` is possible?

zarov replied:
It can, but only for `this.style`. If you take a look at `View`, you have the following line:
```js
labelLayer.style = layer.style;
```
which default to `{}` if you check `ColorLayer`.

gchoqueux commented on src/Layer/LabelLayer.js:
```diff
@@ -0,0 +1,221 @@
+
+                // FIXME: Style management needs improvement, see #1318.
+                const label = new Label(content,
+                    coord.clone(),
```
If is possible, could you remove `clone()`?

zarov replied:
Well `coord` is global, so I can't really avoid this.

gchoqueux replied:
ok

gchoqueux reviewed (approved):
**Thank you very much for this big job that brings an important feature!**

Mkonini commented:
Good job, thnx a lot!
